ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/API.hs

```diff
@@ -0,0 +1,239 @@
+-- database.
+data ImmutableDB m = ImmutableDB
+  { -- | Close the database.
+    --
+    -- Idempotent.
+    --
+    -- __Note__: Use 'withDB' instead of this function.
+    closeDB
+      :: HasCallStack => m (Either ImmutableDBError ())
+      -- TODO remove this operation from the public API, replace it with a
+      -- reopen + validate function that can be used in case of an error.
+      --
+      -- The idea is that the 'ImmutableDB' handle/record can be stored in a
+      -- Reader (instead of State), so in case of an error, we can reopen the
+      -- same database (after validation) instead of having to create a new
+      -- 'ImmutableDB' record for the same on-disk database.
+      --
+      -- We should still expose this field in an internal record so it can be
+      -- used by 'withDB'.
+
+    -- | Return 'True' when the database is open.
+  , isOpen
+      :: HasCallStack => m Bool
+
+    -- | Return the next free 'RelativeSlot' in the current epoch as an
+    -- 'EpochSlot'.
+    --
+    -- Throws a 'ClosedDBError' if the database is closed.
+  , getNextEpochSlot
+      :: HasCallStack => m (Either ImmutableDBError EpochSlot)
+
+    -- | Return the folder in which the database was opened.
+  , getDBFolder :: FsPath
+
+    -- | Get the binary blob stored at the given 'EpochSlot'.
+    --
+    -- Returns 'Nothing' if no blob was stored at the given (valid) slot.
+    --
+    -- Throws a 'ReadFutureSlotError' if the requested slot is in the future,
+    -- i.e >= 'getNextEpochSlot'.
+    --
+    -- Throws a 'SlotGreaterThanEpochSizeError' if the requested (relative)
+    -- slot was exceeded the size of the corresponding epoch.
+    --
+    -- Throws a 'MissingFileError' if the epoch or index file corresponding
+    -- with the requested epoch was missing from disk.
+    --
+    -- Throws a 'ClosedDBError' if the database is closed.
+  , getBinaryBlob
+      :: HasCallStack
+      => EpochSlot -> m (Either ImmutableDBError (Maybe ByteString))
+
+    -- | Appends a binary blob at the given relative slot in the current epoch
+    -- file.
+    --
+    -- Throws an 'AppendToSlotInThePastError' if the given relative slot is
+    -- before the one returned by 'getNextEpochSlot'.
+    --
+    -- Throws a 'SlotGreaterThanEpochSizeError' if the relative slot exceeds
+    -- the size of the current epoch.
+    --
+    -- If a 'FileSystemError' is encountered while appending, the epoch and
+    -- index files are truncated to the last valid state, the state before the
+    -- write.
+    --
+    -- Throws a 'ClosedDBError' if the database is closed.
+    --
+    -- TODO the given binary blob may not be empty.
+  , appendBinaryBlob
+      :: HasCallStack
+      => RelativeSlot -> Builder -> m (Either ImmutableDBError ())
+
+    -- | Close the current epoch and start a new epoch (the epoch number is
+    -- incremented by 1). The new epoch number is returned.
+    --
+    -- The size of the new epoch must be passed.
```

Comment:
I think the size of epochs is a constant for a given chain, determined by `k`, and so cannot be changed except by a hard fork. But I might be wrong.

Reply:
This decision comes from Edsko and Duncan.